pkg/apb/watch_pod.go

```diff
@@ -18,55 +18,76 @@ package apb
 
 import (
 	"fmt"
-	"time"
 
-	"github.com/openshift/ansible-service-broker/pkg/clients"
+	"reflect"
 
 	apiv1 "k8s.io/api/core/v1"
+	meta_v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/watch"
+	"k8s.io/client-go/kubernetes/typed/core/v1"
 )
 
 var (
 	// ErrorPodPullErr - Error indicating we could not pull the image.
 	ErrorPodPullErr = fmt.Errorf("Unable to pull APB image from it's registry. Please contact your cluster admin")
 )
 
-func watchPod(podName string, namespace string) error {
+func watchPod(podName string, namespace string, podClient v1.PodInterface, statusUpdates chan<- JobState) error {
 	log.Debugf(
 		"Watching pod [ %s ] in namespace [ %s ] for completion",
 		podName,
 		namespace,
 	)
 
-	k8scli, err := clients.Kubernetes()
+	w, err := podClient.Watch(meta_v1.ListOptions{})
 	if err != nil {
-		return fmt.Errorf("Unable to retrive kubernetes client - %v", err)
+		return fmt.Errorf("failed to watch pod %s in namespace %s error: %v", podName, namespace, err)
 	}
-
-	for r := 1; r <= apbWatchRetries; r++ {
-		log.Info("Watch pod [ %s ] tick %d", podName, r)
-
-		podStatus, err := k8scli.GetPodStatus(podName, namespace)
-		if err != nil {
-			return err
+	var send = func(state JobState) {
+		if nil == statusUpdates {
+			return
+		}
+		statusUpdates <- state
+	}
+	for podEvent := range w.ResultChan() {
+		pod, ok := podEvent.Object.(*apiv1.Pod)
+		if !ok {
+			log.Errorf("watch did not return a apiv1.Pod instead returned %v", reflect.TypeOf(podEvent.Object))
+			continue
+		}
+		if pod.Name != podName {
+			log.Debugf("watching pods in namespace %s ignoring pod %s as it is not the pod we are looking for", namespace, pod.Name)
+			continue
 		}
 
+		state := JobState{State: StateInProgress, Podname: podName}
+		lastOp := pod.Annotations["apb_last_operation"]
+		if lastOp != "" {
+			state.Description = lastOp
+		}
+		send(state)
+		podStatus := pod.Status
+		log.Debugf("pod [%s] in phase %s", podName, podStatus.Phase)
 		switch podStatus.Phase {
 		case apiv1.PodFailed:
 			if errorPullingImage(podStatus.ContainerStatuses) {
 				return ErrorPodPullErr
 			}
-			return fmt.Errorf("Pod [ %s ] failed - %v", podName, podStatus.Message)
+			w.Stop()
+			return fmt.Errorf("pod [ %s ] failed - %v", podName, podStatus.Message)
 		case apiv1.PodSucceeded:
+			w.Stop()
 			log.Debugf("Pod [ %s ] completed", podName)
 			return nil
 		default:
 			log.Debugf("Pod [ %s ] %s", podName, podStatus.Phase)
 		}
-
-		time.Sleep(time.Duration(apbWatchInterval) * time.Second)
+		if podEvent.Type == watch.Deleted {
+			w.Stop()
```

Comment:
Yes I think you are correct. In most cases the pod would succeed or fail. So it it was somehow deleted it would likely be an error and it should be reported